### PR TITLE
Add support for Google trusted publishing

### LINKED ISSUE #13551: Trusted publishing: support for Google Cloud service accounts
Support for these would allow PyPI users with Google Cloud-based publishing workflows to benefit from trusted publishing.

An example claim set from a Google Cloud service account, lightly anonymized:

```json
{
  "aud": "{SOME-AUDIENCE}",
  "azp": "{SOME-AZP}",
  "email": "{PROJECT-NUMBER}-compute@developer.gserviceaccount.com",
  "email_verified": true,
  "exp": 1682967007,
  "google": {
    "compute_engine": {
      "instance_creation_timestamp": 1674546966,
      "instance_id": "{INSTANCE-ID}",
      "instance_name": "dev",
      "project_id": "{PROJECT-ID}",
      "project_number": {PROJECT-NUMBER},
      "zone": "us-central1-a"
    }
  },
  "iat": 1682963407,
  "iss": "https://accounts.google.com",
  "sub": "{SOME-AZP}"
}
```

I've tried to keep the substitution names consistent above, to show where field values are duplicated.

Based on that claim set, it looks like the relevant uniquely identifying fields are:

1. `aud` (which should be `pypi`, similar to GitHub-issued JWTs)
2. `azp`: no idea what this is
3. `google.project_id`: presumably configured by a user
4. `google.project_number`: presumably a unique ID that prevents resurrection of `google.project_id`
5. `email`: presumably derivable consistently from `google.project_number`

So, my first educated guess is that we'll want to allow users to configure (3) and (4). Does that sound right to you @di?

---

- [x] Refactor `warehouse.oidc.models` to make adding new publisher models simpler (#13553)
- [x] Add OIDC provider models and services for Google Cloud
    - [x] Models: #13559
    - [x] Services https://github.com/pypi/warehouse/pull/13569
- [x] Support multiple publishers in the UI: https://github.com/pypi/warehouse/pull/13571
- [x] Expose Google Cloud providers through forms and views: https://github.com/pypi/warehouse/pull/15144
- [x] Update emails, etc. to be generic over OIDC provider kinds (xref #13667): https://github.com/pypi/warehouse/pull/13872
- [ ] Update user docs: https://github.com/pypi/warehouse/pull/15192

### PR DESCRIPTION
**Pull request by di**
~~Draft while we get https://github.com/pypi/warehouse/pull/15143 and https://github.com/pypi/warehouse/pull/14063 https://github.com/pypi/warehouse/pull/15148 merged.~~

~~On top of #15143, 09812df38d5d826f43cf6a07422a9d26f055e5fe is the relevant commit.~~

Still needs:
- [x] Tests for the PendingPublisher
- [x] Tests for the Publisher

---

Towards #13551.

The flag is disabled for now while we get docs added.

### CODE REVIEW

**Comment by di** [comment]
Looks like https://squidfunk.github.io/mkdocs-material/reference/content-tabs/#linked-content-tabs-feature-enabled might be the answer for the doc updates needed here.

**Comment by woodruffw** [comment]
> Looks like https://squidfunk.github.io/mkdocs-material/reference/content-tabs/#linked-content-tabs-feature-enabled might be the answer for the doc updates needed here.

I'll do a separate PR for this today!

**Comment by di** [comment]
OK, I think this is good to go, we can queue up docs for this in https://github.com/pypi/warehouse/pull/15173.

File: warehouse/manage/views/__init__.py [hunk 1]
```diff
@@ -1299,6 +1304,115 @@ def add_github_oidc_publisher(self):
...
+        request_method="POST",
+        request_param=GooglePublisherForm.__params__,
+    )
+    def add_google_oidc_publisher(self):
+        if self.request.flags.disallow_oidc(AdminFlagValue.DISALLOW_GOOGLE_OIDC):
+            self.request.session.flash(
+                self.request._(
+                    "Google-based trusted publishing is temporarily disabled. "
+                    "See https://pypi.org/help#admin-intervention for details."
+                ),
```
[the remark this answers is not in the document]
Comment (di) [reply warehouse/manage/views/__init__.py:1317]: Partly to minimize user confusion, but also my plan is to use this as a feature flag until we're ready to enable the publisher.

[the remark this answers is not in the document]
Comment (woodruffw) [reply warehouse/manage/views/__init__.py:1317]: > Is it just a matter of minimizing user confusion? (If OIDC is disabled, they might be confused if they successfully set up a publisher which then doesn't work)
> Or can adding publishers also be problematic during high spam periods?

At least for the GitHub provider, a little bit of both: allowing registration at a time when actually using the publisher is impossible is likely (IMO) to cause user confusion, and _also_ registering the publisher (in GitHub's case) involves making some GitHub API calls that we probably want to disable during periods of spam. The latter is not true for the Google publisher, though 🙂

**Review by woodruffw — APPROVED** [review]
LGTM!

File: warehouse/oidc/forms/google.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+class GooglePublisherBase(forms.Form):
+    __params__ = ["email", "sub"]
+
+    email = wtforms.fields.EmailField(
+        validators=[
+            wtforms.validators.InputRequired(),
+            wtforms.validators.Regexp(
+                r".+@.+\..+", message=_("The email address isn't valid. Try again.")
+            ),
```
Comment (miketheman) [inline warehouse/oidc/forms/google.py:33]: I realize this is probably copy&pasta from `NewEmailMixin`, but could this be replaced with a built-in validator? https://wtforms.readthedocs.io/en/3.0.x/validators/#wtforms.validators.Email

We have the required packages already, and use the validator elsewhere in the codebase.

```suggestion
            wtforms.validators.Email(),
```

**Review by miketheman — APPROVED** [review]
I didn't go through with a fine-tooth comb, but a cursory skim looks generally okay.

I had one validator-based question inline, feel free to take it or leave it - it can always be reevaluated later.

File: warehouse/oidc/forms/google.py (hunk 2 above)
Comment (di) [reply warehouse/oidc/forms/google.py:33]: Aha, I think this is a holdover from before we started using that (this PR is quite old). Nice catch.

**Pull request merged**